core/src/main/java/hudson/util/AtomicFileWriter.java

```diff
@@ -108,35 +139,77 @@ public void close() throws IOException {
      * the {@link #commit()} is called, to simplify coding.
      */
     public void abort() throws IOException {
-        close();
-        tmpFile.delete();
+        closeAndDeleteTempFile();
     }
 
     public void commit() throws IOException {
         close();
-        if (destFile.exists()) {
+        try {
+            // Try to make an atomic move.
+            Files.move(tmpPath, destPath, StandardCopyOption.ATOMIC_MOVE);
+        } catch (IOException e) {
+            // If it falls here that can mean many things. Either that the atomic move is not supported,
+            // or something wrong happened. Anyway, let's try to be over-diagnosing
+            if (e instanceof AtomicMoveNotSupportedException) {
+                LOGGER.log(Level.WARNING, "Atomic move not supported. falling back to non-atomic move.", e);
+            } else {
+                LOGGER.log(Level.WARNING, "Unable to move atomically, falling back to non-atomic move.", e);
+            }
+
+            if (destPath.toFile().exists()) {
+                LOGGER.log(Level.WARNING, "The target file {0} was already existing?!?", destPath);
```

Comment:
How is this an exceptional case? I think the most common use case is `destPath.toFile().exists() == true`, isn't it?

Reply:
@amuniz no, it should not be common actually. See https://github.com/jenkinsci/jenkins/pull/2548#discussion_r84765531

If the underlying FS supports atomic move, it will anyway also replace the file, so we wouldn't even enter that catch.

Now I retrospectively agree the "?!?" is probably a bit too much, though I still want to keep that log overall for more diagnosability.

Reply:
NIT: [`Files.exists(destPath)`](https://docs.oracle.com/javase/8/docs/api/java/nio/file/Files.html#exists-java.nio.file.Path-java.nio.file.LinkOption...-)

Reply:
There seems to be a bug in the JDK about the performances of that method, fixed only in JDK9, that's why I used the old variant.
https://bugs.openjdk.java.net/browse/JDK-8154077
https://bugs.openjdk.java.net/browse/JDK-8153414

It was reported to me by SonarLint I use in my IDE: https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3725